File: client (mobile)/lib/BL/places_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:ump_student_grab_mobile/Model/place_result.dart';
import 'package:ump_student_grab_mobile/Model/place_suggestion.dart';
import 'package:ump_student_grab_mobile/util/shared_preferences_util.dart';

class PlacesService {
  final String _appDomain = dotenv.get("APP_DOMAIN");
  final String _appPort = dotenv.get("APP_PORT");
  final String _apiKey = dotenv.get("API_KEY");

  String get _baseUrl => "http://$_appDomain:$_appPort/api/places";

  Map<String, String> get _headers => {
        "Content-Type": "application/json",
        "X-Api-Key": _apiKey,
      };

  Future<List<PlaceSuggestion>> autocomplete(String query, {String region = "MY"}) async {
    final url = Uri.parse("$_baseUrl/autocomplete?query=${Uri.encodeComponent(query)}&region=$region");
    final response = await http.get(url, headers: _headers);

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body)['data'];
      return data.map((e) => PlaceSuggestion.fromJson(e)).toList();
    }
    return [];
  }

  Future<List<PlaceResult>> searchByText(String query, {String region = "MY"}) async {
    final url = Uri.parse("$_baseUrl/search?query=${Uri.encodeComponent(query)}&region=$region");
    final response = await http.get(url, headers: _headers);

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body)['data'];
      return data.map((e) => PlaceResult.fromJson(e)).toList();
    }
    return [];
  }

  Future<String?> getDirections(String origin, String destination) async {
    final url = Uri.parse(
        "$_baseUrl/directions?origin=${Uri.encodeComponent(origin)}&destination=${Uri.encodeComponent(destination)}");
    final response = await http.get(url, headers: _headers);

    if (response.statusCode == 200) {
      return jsonDecode(response.body)['data']['encodedPolyline'] as String?;
    }
    return null;
  }
}

File: server/src/main/java/com/ump/studentgrab/application/service/PlacesService.java
package com.ump.studentgrab.application.service;

import com.ump.studentgrab.application.dto.response.DirectionsResponse;
import com.ump.studentgrab.application.dto.response.PlaceSearchResponse;
import com.ump.studentgrab.application.dto.response.PlaceSuggestionResponse;
import com.ump.studentgrab.application.exception.ResourceNotFoundException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.List;
import java.util.Map;

@Service
public class PlacesService {

    private static final String PLACES_BASE_URL = "https://maps.googleapis.com/maps/api/place";
    private static final String DIRECTIONS_BASE_URL = "https://maps.googleapis.com/maps/api/directions";

    private final String apiKey;
    private final RestClient restClient;

    public PlacesService(@Value("${google.maps.api-key}") String apiKey) {
        this.apiKey = apiKey;
        this.restClient = RestClient.create();
    }

    public List<PlaceSuggestionResponse> autocomplete(String query, String region) {
        String url = UriComponentsBuilder.fromHttpUrl(PLACES_BASE_URL + "/autocomplete/json")
                .queryParam("input", query)
                .queryParam("region", region)
                .queryParam("key", apiKey)
                .toUriString();

        Map<?, ?> body = restClient.get().uri(url).retrieve().body(Map.class);
        List<?> predictions = (List<?>) body.get("predictions");

        if (predictions == null || predictions.isEmpty()) return List.of();

        return predictions.stream()
                .map(p -> (Map<?, ?>) p)
                .map(p -> new PlaceSuggestionResponse(
                        (String) p.get("description"),
                        (String) p.get("place_id")
                ))
                .toList();
    }

    public List<PlaceSearchResponse> searchByText(String query, String region) {
        String url = UriComponentsBuilder.fromHttpUrl(PLACES_BASE_URL + "/textsearch/json")
                .queryParam("query", query)
                .queryParam("region", region)
                .queryParam("key", apiKey)
                .toUriString();

        Map<?, ?> body = restClient.get().uri(url).retrieve().body(Map.class);
        List<?> results = (List<?>) body.get("results");

        if (results == null || results.isEmpty()) return List.of();

        return results.stream()
                .map(r -> (Map<?, ?>) r)
                .filter(r -> r.get("geometry") instanceof Map<?, ?> g
                        && g.get("location") instanceof Map<?, ?>)
                .map(r -> {
                    Map<?, ?> geometry = (Map<?, ?>) r.get("geometry");
                    Map<?, ?> location = (Map<?, ?>) geometry.get("location");
                    double lat = ((Number) location.get("lat")).doubleValue();
                    double lng = ((Number) location.get("lng")).doubleValue();
                    return new PlaceSearchResponse(
                            (String) r.get("name"),
                            (String) r.get("place_id"),
                            lat,
                            lng
                    );
                })
                .toList();
    }

    public DirectionsResponse getDirections(String origin, String destination) {
        String url = UriComponentsBuilder.fromHttpUrl(DIRECTIONS_BASE_URL + "/json")
                .queryParam("origin", origin)
                .queryParam("destination", destination)
                .queryParam("key", apiKey)
                .toUriString();

        Map<?, ?> body = restClient.get().uri(url).retrieve().body(Map.class);
        List<?> routes = (List<?>) body.get("routes");

        if (routes == null || routes.isEmpty()) {
            throw new ResourceNotFoundException("No route found between the given locations");
        }

        Map<?, ?> route = (Map<?, ?>) routes.get(0);
        Map<?, ?> overviewPolyline = route.get("overview_polyline") instanceof Map<?, ?> p ? p : null;

        if (overviewPolyline == null || overviewPolyline.get("points") == null) {
            throw new ResourceNotFoundException("Route found but polyline data is missing");
        }

        String encodedPolyline = (String) overviewPolyline.get("points");
        return new DirectionsResponse(encodedPolyline);
    }
}
